### ISSUE TITLE
Motorola Basic Pricacy

### ISSUE THREAD

**Issue by iz4tow** [lwvmobile/dsd-fme#161]
Hi,
if you are interested I can generate and provide you wav recorded files with encrypted Motorola/DMR with the key.
Let me know
Thx

**Comment by lwvmobile** [comment]
Oh, I can't say I particularly need any encrypted samples at the moment, but thanks for the offer.

**Comment by iz4tow** [comment]
Ok.
Anyway to do some test...how can I use dsd-fme with the known key?

I mean...On Motorola CPS 2.0 I set the privacy key = TestPassword! but on dsd-fme the options accept only hex.

**Comment by billghunt** [comment]
Motorola's Basic Privacy only has a choice of 1-256, as no input is allowed.  You must mean Enhanced Privacy, because it's a 40 bit hex key (RC40).  I can't imagine Motorola would allow you put non-hex characters as a key.  I hope I'm wrong, but it sounds like you're using a hacked version of CPS.

**Comment by iz4tow** [comment]
I'm using CPS 2.134.76.0
Here the screenshot

![image](https://github.com/lwvmobile/dsd-fme/assets/33699109/2e3e4aca-5bf0-471f-8afb-43305d799f6a)

**Comment by lwvmobile** [comment]
You wan to set the key under the Privacy heading on that page, and not set the RAS information. The RAS is used as a method of 'protection' to keep radios off of other repeaters unless you know the password. I use the word 'protection' loosely, since all it effectively does is scramble the CRC values on selected items.

![262789798-2e3e4aca-5bf0-471f-8afb-43305d799f6a](https://github.com/lwvmobile/dsd-fme/assets/59371473/bcf5215e-1ae6-457a-9ee2-c49b4a6259c2)

**Comment by iz4tow** [comment]
Ok you are right. I tried with -b 1 and it works!
So for enhanced privacy should I try with -1 or -K, am I right?
Thanks!

**Comment by iz4tow** [comment]
and another question, I tried with `dsd-fme -k` and this file:
```
id,key
1,1F00
2,E300
3,FC00
4,2503
5,3A03
6,C603
7,D903
8,4A05
9,5505
10,A905
11,B605
12,6F06
13,7006
14,8C06
15,9306
16,2618
17,3918
18,C518
19,DA18
20,031B
21,1C1B
22,E01B
23,FF1B
24,6C1D
25,731D
26,8F1D
27,901D
28,491E
29,561E
30,AA1E
31,B51E
32,4B28
33,5428
34,A828
35,B728
36,6E2B
37,712B
38,8D2B
39,922B
40,012D
41,1E2D
42,E22D
43,FD2D
44,242E
45,3B2E
46,C72E
47,D82E
48,6D30
49,7230
50,8E30
51,9130
52,4833
53,5733
54,AB33
55,B433
56,2735
57,3835
58,C435
59,DB35
60,0236
61,1D36
62,E136
63,FE36
64,2B49
65,3449
66,C849
67,D749
68,0E4A
69,114A
70,ED4A
71,F24A
72,614C
73,AE4C
74,824C
75,9D4C
76,444F
77,5B4F
78,A74F
79,B84F
80,0D51
81,1251
82,EE51
83,F151
84,2852
85,3752
86,CB52
87,D452
88,4754
89,5854
90,A454
91,BB54
92,6257
93,7D57
94,8157
95,9E57
96,6061
97,7F61
98,8361
99,9C61
100,4562
101,5A62
102,A662
103,B962
104,2A64
105,3564
106,C964
107,D664
108,0F67
109,1067
110,EC67
111,F367
112,4679
113,5979
114,A579
115,BA79
116,637A
117,7C7A
118,807A
119,9F7A
120,0C7C
121,137C
122,EF7C
123,F07C
124,297F
125,367F
126,CA7F
127,D57F
128,4D89
129,5289
130,AE89
131,B189
132,688A
133,778A
134,8B8A
135,948A
136,078C
137,188C
138,E48C
139,FB8C
140,228F
141,3D8F
142,C18F
143,DE8F
144,6B91
145,7491
146,8891
147,9791
148,4E92
149,5192
150,AD92
151,B292
152,2194
153,3E94
154,C294
155,DD94
156,0497
157,1B97
158,E797
159,F897
160,06A1
161,19A1
162,E5A1
163,FAA1
164,23A2
165,3CA2
166,C0A2
167,DFA2
168,4CA4
169,53A4
170,AFA4
171,B0A4
172,69A7
173,76A7
174,8AA7
175,95A7
176,20B9
177,3FB9
178,C3B9
179,DCB9
180,05BA
181,1ABA
182,E6BA
183,F9BA
184,6ABC
185,75BC
186,89BC
187,96BC
188,4FBF
189,50BF
190,ACBF
191,B3BF
192,66C0
193,79C0
194,85C0
195,9AC0
196,43C3
197,5CC3
198,A0C3
199,BFC3
200,2CC5
201,33C5
202,CFC5
203,D0C5
204,09C6
205,16C6
206,EAC6
207,F5C6
208,84D0
209,85DF
210,8AD3
211,8BDC
212,B6D5
213,B7DA
214,B8D6
215,B9D9
216,D0DA
217,D1D5
218,DED9
219,DFD6
220,E2DF
221,E3D0
222,ECDC
223,EDD3
224,2DE8
225,32E8
226,CEE8
227,D1E8
228,08EB
229,17EB
230,EBEB
231,F4EB
232,67ED
233,78ED
234,84ED
235,9BED
236,42EE
237,5DEE
238,A1EE
239,BEEE
240,0BF0
241,14F0
242,E8F0
243,F7F0
244,2EF3
245,31F3
246,CDF3
247,D2F3
248,41F5
249,5EF5
250,A2F5
251,BDF5
252,64F6
253,7BF6
254,87F6
255,98F6
```
but is not working, I also tried to convert the keys into dec...but still not working.
```
id,key
1,7936
2,58112
3,64512
4,9475
5,14851
6,50691
7,55555
8,18949
9,21765
10,43269
11,46597
12,28422
13,28678
14,35846
15,37638
16,9752
17,14616
18,50456
19,55832
20,795
21,7195
22,57371
23,65307
24,27677
25,29469
26,36637
27,36893
28,18718
29,22046
30,43550
31,46366
32,19240
33,21544
34,43048
35,46888
36,28203
37,28971
38,36139
39,37419
40,301
41,7725
42,57901
43,64813
44,9262
45,15150
46,50990
47,55342
48,27952
49,29232
50,36400
51,37168
52,18483
53,22323
54,43827
55,46131
56,10037
57,14389
58,50229
59,56117
60,566
61,7478
62,57654
63,65078
64,11081
65,13385
66,51273
67,55113
68,3658
69,4426
70,60746
71,62026
72,24908
73,44620
74,33356
75,40268
76,17487
77,23375
78,42831
79,47183
80,3409
81,4689
82,61009
83,61777
84,10322
85,14162
86,52050
87,54354
88,18260
89,22612
90,42068
91,47956
92,25175
93,32087
94,33111
95,40535
96,24673
97,32609
98,33633
99,40033
100,17762
101,23138
102,42594
103,47458
104,10852
105,13668
106,51556
107,54884
108,3943
109,4199
110,60519
111,62311
112,18041
113,22905
114,42361
115,47737
116,25466
117,31866
118,32890
119,40826
120,3196
121,4988
122,61308
123,61564
124,10623
125,13951
126,51839
127,54655
128,19849
129,21129
130,44681
131,45449
132,26762
133,30602
134,35722
135,38026
136,1932
137,6284
138,58508
139,64396
140,8847
141,15759
142,49551
143,56975
144,27537
145,29841
146,34961
147,38801
148,20114
149,20882
150,44434
151,45714
152,8596
153,16020
154,49812
155,56724
156,1175
157,7063
158,59287
159,63639
160,1697
161,6561
162,58785
163,64161
164,9122
165,15522
166,49314
167,57250
168,19620
169,21412
170,44964
171,45220
172,27047
173,30375
174,35495
175,38311
176,8377
177,16313
178,50105
179,56505
180,1466
181,6842
182,59066
183,63930
184,27324
185,30140
186,35260
187,38588
188,20415
189,20671
190,44223
191,46015
192,26304
193,31168
194,34240
195,39616
196,17347
197,23747
198,41155
199,49091
200,11461
201,13253
202,53189
203,53445
204,2502
205,5830
206,60102
207,62918
208,34000
209,34271
210,35539
211,35804
212,46805
213,47066
214,47318
215,47577
216,53466
217,53717
218,57049
219,57302
220,58079
221,58320
222,60636
223,60883
224,11752
225,13032
226,52968
227,53736
228,2283
229,6123
230,60395
231,62699
232,26605
233,30957
234,34029
235,39917
236,17134
237,24046
238,41454
239,48878
240,3056
241,5360
242,59632
243,63472
244,12019
245,12787
246,52723
247,54003
248,16885
249,24309
250,41717
251,48629
252,25846
253,31734
254,34806
255,39158
```

**Comment by iScottybotty** [comment]
Buddy, that’s RAS (Restricted Access to Network) and not BP nor EP. 

RAS is not an encryption per say, it just stops voice being passed ove the air without the correct authentication, key. 

RAS however does NOT stop the listening IF using DSD-FME or a scanner, it really only applies to transceivers. 

For encryption, you need to click the Security tab within the CPS. 

Easy mistake to make, but again, your using RAS ability and not encryption. 

Hope this helps.

**Comment by iz4tow** [comment]
@iScottybotty. Yes....but if I would to build a keyfile for basic privacy?

**Comment by iScottybotty** [comment]
I understand. Unfortunately I don’t know the answer to this one I’m afraid. I personally tend to enter them manually. 👍

**Comment by lwvmobile** [comment]
You don't need to import Moto BP keys like that, those values are already in the code. the -K and the -k options are for importing RC4 keys, Hytera BP keys, NXDN scrambler keys, etc. In the case of keys that don't have, or don't always have a key id, you can load a TG value and a key value. 

For example, if you wanted to import multiple Moto BP or NXDN keys to load in per TG, you would set up a csv file as such

```
TG(dec), Key ID(dec)
100,1
200,2
300,3
```

If you were loading up multiple RC4 keys, you would make a file like this:

```
KeyID(hex), Key Value(hex)
1, 9DEADBEEF9
C7, 420B00B1E5
```

If you are only going to test a single key value, its probably a lot easier to just specify a single key with the appropriate switch.

**Comment by iz4tow** [comment]
Ok. Thanks. Where can I find in the code the BP key list?

**Comment by lwvmobile** [comment]
https://github.com/lwvmobile/dsd-fme/blob/79339d087d082512724a69c96f2707fc8b933efa/src/dsd_mbe.c#L71

**Comment by iz4tow** [comment]
> You don't need to import Moto BP keys like that, those values are already in the code. the -K and the -k options are for importing RC4 keys, Hytera BP keys, NXDN scrambler keys, etc. In the case of keys that don't have, or don't always have a key id, you can load a TG value and a key value.
> 
> For example, if you wanted to import multiple Moto BP or NXDN Scrambler keys to load in per TG, you would set up a csv file as such
> 
> ```
> TG(dec), Key ID(dec)
> 100,1
> 200,2
> 300,3
> ```
> 
> If you were loading up multiple RC4 keys, you would make a file like this:
> 
> ```
> KeyID(hex), Key Value(hex)
> 1, 9DEADBEEF9
> C7, 420B00B1E5
> ```
> 
> If you are only going to test a single key value, its probably a lot easier to just specify a single key with the appropriate switch.

I tried this...but it worked from 1 to 10....I'd like to have all 255 keys loaded in a single session from file

**Comment by lwvmobile** [comment]
You don't have to load all 255 Moto BP keys, they are already available, you just have to tell dsd-fme which key id you want to use, and it'll pull the correct value from that table. The point of the ` -k key.csv` switch when using Moto BP is to link a TG value to a certain BP key id, since basic privacy does not inherently have a key id broadcast over the air, and its an easy way to switch multiple BP keys on the fly.  

Again, you don't have to load in what's already there. I'm honestly not even sure what you're trying to accomplish right now.

**Comment by iz4tow** [comment]
I have 2 goals:
1 - use all 255 MOto BP in a single session
2 - load keys from file.

I created a file like this:

```
100,1
200,2 
[...]
25500,255
```

And tried with my Motorola DMR and dsd-fme -k keys.csv
Results:
- keys 1 to 9 -> works
- keys 10 to 255 -> doesn't work

**Comment by lwvmobile** [comment]
So, what's your Radio's TG ID? Is it static, or are you changing it to be 100, 200, ..., 25500?

**Comment by iz4tow** [comment]
I'm sorry I didn't explain very well.

I have 3 different "radio network" on the same frequency: 
- network 1 with key 1, so 1F00
- network 2 with key 2, so E300
- network 3 with key 3, so FC00

and I'd like to listen all from a single PC.

So how can I?

**Comment by lwvmobile** [comment]
The only way you'll be able to do all of them on the same frequency/same instance of DSD-FME is if they all have a unique but uniform TGID and that TGID uses the same key. If its Unit to Unit calls, or private calls, then it gets a lot trickier. Keep in mind, Moto BP does not issue key id values over the air, this is just a quick 'experimental' bit of code I wrote to load in keys based on TG value, it can't (nor can I) predict and write code for countless different scenarios of how other systems or combinations of systems are set up.

So, in other words, I really don't know how to make it work in your scenario, I still don't truly understand your scenario exactly, even if I did, It might not be possible to do all of them if the keys aren't different on a per talkgroup level.
